File: solutions/haskell/isbn-verifier/1/src/IsbnVerifier.hs
module IsbnVerifier (isbn) where
import Data.Char (isDigit, digitToInt)

isbn :: String -> Bool
isbn xs = length digits == 10 && controlSum `mod` 11 == 0
    where
      isbn2ints ('X' : ys) = 10 : map digitToInt (filter isDigit ys)
      isbn2ints ys = map digitToInt (filter isDigit ys)
      digits = isbn2ints (reverse xs)
      controlSum = sum (zipWith (*) digits [1..])

File: solutions/erlang/isbn-verifier/1/src/isbn_verifier.erl
-module(isbn_verifier).
-export([is_valid/1]).
-import(lists, [sum/1, map/2, reverse/1, seq/2, zip/2]).

is_valid(Isbn) ->
  X2 = zip([X - $0 || X <- Isbn, is_digit(X) ], reverse(seq(1,10))),
  sum([X * Y || {X,Y} <- X2]) rem 11 == 0.

is_digit(C) -> (C >= $0) andalso (C =< $9).

